compressors/shannon_fano_elias_coder.py

```diff
@@ -0,0 +1,94 @@
+"""Shannon Fano Elias Coding
+Shannon Fano Elias coding is a prefix free coding, and is a precursor to Arithmetic coding
+
+Encoding can be seen from here: https://en.wikipedia.org/wiki/Shannon%E2%80%93Fano%E2%80%93Elias_coding
+NOTE: As Shannon-Fano-Elias is a prefix-free code, one can compute the codewords once and use that for encoding.
+The decoding can also employ the prefix-free tree. We however have implemented both encoding/decoding in a way 
+which motivates Arithmetic coding 
+
+"""
+
+from utils.bitarray_utils import float_to_bitarrays, BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeEncoder, PrefixFreeDecoder
+from core.prob_dist import ProbabilityDist
+import math
+import numpy as np
+
+
+class ShannonFanoEliasEncoder(PrefixFreeEncoder):
+    def __init__(self, prob_dist: ProbabilityDist):
+        self.prob_dist = prob_dist
+
+        # compute a dictionary holding cumulative probabilities
+        self.cumulative_prob_dict = self.prob_dist.cumulative_prob_dict
```

Comment:
is this a fn call to compute the cumprob?

Reply:
yes it is a function call.. but if we decorate it with `@property`, it can be viewed as an attribute + it gets cached for efficiency

Reply:
i will add a comment with that effect in the prob_dist code